idxtype: add idxtype package

andy-kimball opened the pull request:
Add new sql/sem/idxtype package that defines an index type enum with values
like forward, inverted, and vector, as well as "capability" functions describing
what the index can or cannot do. Code across SQL needs to check the index
type and previously it used multiple different enums and bools. This commit
continues the process of unifying these mechanisms (future commits will do
even more).

Epic: CRDB-42943

Release note: None

andy-kimball committed:
idxtype: add idxtype package

Add new sql/sem/idxtype package that defines an index type enum with values
like forward, inverted, and vector. Code across SQL needs to check the index
type and previously it used multiple different enums and bools. This commit
continues the process of unifying these mechanisms (future commits will do
even more).

Epic: CRDB-42943

Release note: None

andy-kimball committed:
opt: convert from IsInverted/IsVector to idxtype.T

Now that we have the new idxtype.T enumeration that specifies the type
of index (e.g. forward, inverted, vector), use that in the opt packages
rather than IsInverted and IsVector boolean functions. This will make
our code more extensible, such that adding new index types is easier,
starting with vector indexes.

Epic: CRDB-42943

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139786)
<!-- Reviewable:end -->

andy-kimball commented:
Note that there are 2 commits in this PR: the 1st commit contains mostly Foundations changes and the 2nd commit mostly Queries changes.

andy-kimball committed:
idxtype: add capability helper functions to idxtype

Add helper functions that specify the capabilities of various index
types. For example, only FORWARD indexes can be primary or unique,
and only INVERTED and VECTOR indexes support prefix columns.

Use these new helper functions across the SQL packages.

Epic: CRDB-42943

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for cleaning things up!

> Reviewed 101 of 101 files at r1, 40 of 40 files at r2, 13 of 13 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-:-OHcEJ739_z1rbDpzmZE:bwv9qnn)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @rafiss)

___
*[`pkg/sql/sem/idxtype/idxtype.proto` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHc9-5G4joyYrCB-ELZ:-OHc9-5G4joyYrCB-EL_:b-ispa0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/63009ad6ef9561e74372d33e3d85a53afac1ba45/pkg/sql/sem/idxtype/idxtype.proto#L21)):*
> ```proto
>   FORWARD = 0;
>   // INVERTED can contain multiple entries for each row in the table, in order
>   // to index collection or composite data types like JSONB, ARRAY, and
> ```

nit: missing a word in "... in order to index collection..."?
___
*[`pkg/sql/sem/idxtype/idxtype.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDb7x8mTIesk6si4l:-OHcDb7x8mTIesk6si4m:b-nb02x9) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/sem/idxtype/idxtype.go#L11)):*
> ```Go
> // contains unique keys sorted according to the primary ordering of the table.
> // Secondary indexes refer to rows in the primary index by unique key value.
> func CanBePrimary(t T) bool {
> ```

I'm curious if you thought about making these methods of `T` instead of functions. It'd make calling them slightly more succinct. Maybe there is an benefit to functions I'm not seeing. Feel free to leave as-is—I don't feel strongly.
___
*[`pkg/sql/opt/optbuilder/groupby.go` line 993 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcCVRo7CjPgrWHFbA6:-OHcCVRo7CjPgrWHFbA7:bs8ytto) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/opt/optbuilder/groupby.go#L993)):*
> ```Go
> 	for i := 1; i < tab.IndexCount(); i++ {
> 		index := tab.Index(i)
> 		if !index.IsUnique() || idxtype.CanBeUnique(index.Type()) {
> ```

I think this should be `!idxtype.CanBeUnique(index.Type())`.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 311 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDG5AAFA_LWq0csT7:-OHcDG5AAFA_LWq0csT8:b-3m9c1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L311)):*
> ```Go
> 	if columnNode.OpClass != "" && (!lastColIdx || !idxtype.SupportsOpClass(n.Type)) {
> 		panic(pgerror.New(pgcode.DatatypeMismatch,
> 			"operator classes are only allowed for the last column of an inverted index"))
> ```

nit: "inverted" should be the index type name like above
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 316 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDUhq7lwAQJ4X3CYO:-OHcDUhq7lwAQJ4X3CYP:b-3m9c1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L316)):*
> ```Go
> 	if idxtype.AllowExplicitDirection(n.Type) && columnNode.Direction == tree.Descending && lastColIdx {
> 		panic(pgerror.New(pgcode.FeatureNotSupported,
> 			"the last column in an inverted index cannot have the DESC option"))
> ```

nit: "inverted" should be the index type name like above
___
*[`pkg/sql/randgen/schema.go` line 264 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcCugS1ryr_sm7B3xC:-OHcCugS1ryr_sm7B3xD:b-qw1gkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/randgen/schema.go#L264)):*
> ```Go
> 			continue
> 		}
> 		if idxtype.CanBePrimary(indexDef.Type) && pk != nil {
> ```

nit: `!idxtype.CanBePrimary(indexDef.Type)`


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/opt/optbuilder/groupby.go` line 993 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcCVRo7CjPgrWHFbA6:-OHcVhfW-SVlf0QMErHz:bh6480w) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/opt/optbuilder/groupby.go#L993)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think this should be `!idxtype.CanBeUnique(index.Type())`.
</blockquote></details>

You're right. This caused a test failure as well, which shows we have good test coverage here.
___
*[`pkg/sql/randgen/schema.go` line 264 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcCugS1ryr_sm7B3xC:-OHcW96sEWolxsOfczG8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/randgen/schema.go#L264)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: `!idxtype.CanBePrimary(indexDef.Type)`
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 311 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDG5AAFA_LWq0csT7:-OHcX-5p6p0lgfGv_N0J:banl32i) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L311)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "inverted" should be the index type name like above
</blockquote></details>

This one is trickier, b/c it uses "an", where vector needs to use "a". In a future PR where I enable vector indexes, I'll look more closely at this and other cases. For now, I changed the error messages to use "an inverted or vector index".
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 316 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDUhq7lwAQJ4X3CYO:-OHcXIiJ9estdAUU9zPN:brhumzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L316)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "inverted" should be the index type name like above
</blockquote></details>

Same here.
___
*[`pkg/sql/sem/idxtype/idxtype.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDb7x8mTIesk6si4l:-OHcXJzK9T0hwGi9A6MD:blgcasm) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/sem/idxtype/idxtype.go#L11)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm curious if you thought about making these methods of `T` instead of functions. It'd make calling them slightly more succinct. Maybe there is an benefit to functions I'm not seeing. Feel free to leave as-is—I don't feel strongly.
</blockquote></details>

Yeah, I went back and forth. I worry a bit about setting the precedent of having complex methods on a simple idxtype enum. While these methods *are* simple right now, I worry that people will add more and more complexity over time, e.g. by adding methods that involve coltypes. I was also thinking about a future where we have a much more elaborate index type abstraction, similar to what PG has. But maybe we never get there, and we should just do what is most ergonomic right now - I went ahead and switched it.
___
*[`pkg/sql/sem/idxtype/idxtype.proto` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHc9-5G4joyYrCB-ELZ:-OHcYxEm8Kb26sB0Hrf8:bxc1o94) ([raw file](https://github.com/cockroachdb/cockroach/blob/63009ad6ef9561e74372d33e3d85a53afac1ba45/pkg/sql/sem/idxtype/idxtype.proto#L21)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: missing a word in "... in order to index collection..."?
</blockquote></details>

I reworded this.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
What do people think about defining the alias `type IndexType = idxtype.T`? While on the one hand it seems to follow the precedent of `tree` defining its own types, it also means that devs will have the cognitive burden of deciding whether to use,  e.g., `idxtype.INVERTED` vs. `tree.IndexTypeInverted`. This is not a hard decision in the `tree` package itself, but it's more subtle in the schema change packages, where both are used. I end up using `tree.IndexType` where I'm interacting with the AST and `idxtype.T` where I'm not, but it's really easy to mix them up. Should we just use `idxtype` directly in the `tree` package?

mgartner reviewed (approved):
I think we should get rid of the alias.

> Reviewed 51 of 51 files at r4, 38 of 38 files at r5, 14 of 16 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-:-OHcbj5r5GDi6f2AHVsV:b-598be2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 311 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDG5AAFA_LWq0csT7:-OHcd5CP5PIqJgXXxS5A:b-81nlyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L311)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This one is trickier, b/c it uses "an", where vector needs to use "a". In a future PR where I enable vector indexes, I'll look more closely at this and other cases. For now, I changed the error messages to use "an inverted or vector index".
</blockquote></details>

SGTM.
___
*[`pkg/sql/sem/idxtype/idxtype.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHcDb7x8mTIesk6si4l:-OHccJ7c5fhNitQvIb0X:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ece48a9338d165aa79045b31505767aaf20b22/pkg/sql/sem/idxtype/idxtype.go#L11)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Yeah, I went back and forth. I worry a bit about setting the precedent of having complex methods on a simple idxtype enum. While these methods *are* simple right now, I worry that people will add more and more complexity over time, e.g. by adding methods that involve coltypes. I was also thinking about a future where we have a much more elaborate index type abstraction, similar to what PG has. But maybe we never get there, and we should just do what is most ergonomic right now - I went ahead and switched it.
</blockquote></details>

:+1: 
___
*[`pkg/sql/sem/idxtype/idxtype.proto` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHc9-5G4joyYrCB-ELZ:-OHccPus9sDeFi8PXdgM:b-jzhcd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/63009ad6ef9561e74372d33e3d85a53afac1ba45/pkg/sql/sem/idxtype/idxtype.proto#L21)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I reworded this.
</blockquote></details>

Ahh I see now that it was grammatically correct the first time, I was confusing "index" as a noun instead of a verb. It's easier to read now though.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
I'm fine with not having the alias. We definitely should be pretty aggressive with the `disallowed_imports` in the `idxtype/BUILD.bazel` file to ensure that it remains a lightweight leaf package.

`tree` already has its own disallowed list. the new package should at least match that, but perhaps be even more aggressive if it can be.

https://github.com/cockroachdb/cockroach/blob/fd09f27e80b294a0e5be10d4f7cc3ee0b193d9ff/pkg/sql/sem/tree/BUILD.bazel#L307-L320

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-:-OHcdVAy-xxb76hlKvuQ:b-y2qmhs)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

andy-kimball committed:
tree: use idxtype package

Update the tree package to use idxtype.T directly rather than using an
alias for it. Add a check to the idxtype BUILD.bazel package that ensures
it stays a leaf package, so that the tree package doesn't take on
additional dependencies beyond what idxtype needs.

Epic: CRDB-42943

Release note: None

andy-kimball commented:
I added a commit that uses `idxtype` directly in `tree`. I also added a `BUILD.bazel` rule to the `idxtype` package that ensures it stays a leaf package.

rafiss reviewed (approved):
thanks for this organizational improvement!

mgartner reviewed (approved):
> Reviewed 1 of 301 files at r14, 3 of 56 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @andy-kimball, @angles-n-daemons, @arjunmahishi, @DarrylWong, @herkolategan, @jbowens, and @wenyihu6)

___
*[`-- commits` line 44 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/139786#-OHe7OMYCWLUFMPA0dSe:-OHe7OMYCWLUFMPA0dSf:bcpbjt4):*
nit: Looks like the leaf package check was included in the previous commit, not this one.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors=rafiss,mgartner

mgartner commented:
bors r=rafiss,mgartner
